README.md
Uclogic-tools
=============

Uclogic-tools is a collection of programs for collecting and analyzing
diagnostic information from UC-Logic graphics tablets (rebranded as Huion,
Yiynova, Ugee, Monoprice, Turcom and others).

Installation
------------

Download one of the release packages from the [releases
page](https://github.com/DIGImend/uclogic-tools/releases).

Use your Linux distribution tools to install either .rpm or .deb packages.

To build uclogic-tools from the source code, you will need the libusb
development package installed. It is usually named `libusb-1.0-0-dev` or
`libusbx-devel`.

If you're using a release tarball, you can build uclogic-tools by executing
this command in the unpacked source code directory:

    ./configure && make

If you're building a development version downloaded with Git or via GitHub's
"Download ZIP" link, you will also need `autoconf` and `automake` installed on
all distros, as well as `pkg-config` on Debian-based systems. After that you
should be able to build uclogic-tools by executing the following in the source
code directory:

    autoreconf -i -f && ./configure && make

After building uclogic-tools from the source code, you can run the compiled
programs directly, or install them with this command, from the same directory:

    sudo make install

Usage
-----

Uclogic-tools contains two utilities: `uclogic-probe` and `uclogic-decode`.

`Uclogic-probe` dumps diagnostics information from UC-Logic (and rebranded)
graphics tablets and attempts to enable additional functionality.

`Uclogic-decode` attempts to extract tablet parameters from the information
dumped by `uclogic-probe`.

Note that the additional functions might be incompatible with the tablet
driver you're currently using and the tablet might stop working properly after
you execute `uclogic-probe`. To fix that simply reconnect the tablet.

`Uclogic-probe` accepts two arguments: bus number and device address. You can
find them in `lsusb` output by looking for a device with vendor ID 256c and
product ID 006e.

For example, in this `lsusb` output:

    Bus 001 Device 003: ID 256c:006e  
    Bus 001 Device 001: ID 1d6b:0001 Linux Foundation 1.1 root hub
    Bus 002 Device 001: ID 1d6b:0001 Linux Foundation 1.1 root hub
    Bus 003 Device 001: ID 1d6b:0001 Linux Foundation 1.1 root hub
    Bus 004 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
    Bus 005 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
    Bus 006 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub

The first line corresponds to a Huion tablet, and so its bus number is 1,
device address is 3 and you probe it like this:

    sudo uclogic-probe 1 3

The output will be something like this:

    M 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
    P 31 00 30 00 35 00 39 00 34 00 00 00 00 00 00 00 00 00 00 00
    S 64 0E 03 40 9C A8 61 03 00 FF 07 A0 0F 08 00
    S 65 04 03 20 A0
    S 6E 04 03 00 30
    S 79 14 03 48 00 41 00 36 00 30 00 2D 00 46 00 34 00 30 00 30 00
    S 7A 08 03 01 08 00 00 00 00
    S 7B 0C 03 48 00 4B 00 20 00 4F 00 6E 00

The above is what a driver developer would need when asking about a
`uclogic-probe` output.

`Uclogic-decode` simply expects `uclogic-probe` output on its input. E.g. if
you saved the output of `uclogic-probe` into a file named "probe.txt", then
this command would decode it:

    uclogic-decode < probe.txt

You can pipe `uclogic-probe` output directly to `uclogic-decode` too:

    sudo uclogic-probe 1 3 | uclogic-decode

For the diagnostics dump above either of these commands will produce this:

      Manufacturer: ????????
           Product: 10594?????
             Max X: 40000
             Max Y: 25000
      Max pressure: 2047
        Resolution: 4000
    Internal model: HA60-F400
    Buttons status: HK On


debian/changelog
uclogic-tools (5) unstable; urgency=low

  * Increase size of the decoding buffer in uclogic-decode to support all
    descriptor sizes.
  * Ignore missing manufacturer/product string descriptors.
  * Add dumping of Huion new-model string descriptors, including another
    one with parameters, which enables the new-model mode.
  * Improve decoded output formatting.
  * Support specifying uclogic-decode input on the command line.

 -- Nikolai Kondrashov <spbnick@gmail.com>  Sat, 01 Dec 2018 17:14:14 +0200

uclogic-tools (4) unstable; urgency=low

  * Rename the package to uclogic-tools.
  * Add enabling abstract button mode/decoding of buttons status.

 -- Nikolai Kondrashov <spbnick@gmail.com>  Sun, 15 Mar 2015 19:50:06 +0200

huion-tools (3) unstable; urgency=low

  * Add dumping/decoding of manufacturer and product.
  * Add internal model name decoding.

 -- Nikolai Kondrashov <spbnick@gmail.com>  Mon, 14 Oct 2013 11:03:29 +0300

huion-tools (2) unstable; urgency=low

  * Initial Release.

 -- Nikolai Kondrashov <spbnick@gmail.com>  Mon, 23 Sep 2013 15:55:36 +0300


huion-probe.c
/*
 * Copyright (C) 2013 Nikolai Kondrashov
 *
 * This file is part of huion-tools.
 *
 * Huion-tools is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Huion-tools is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with huion-tools; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 * @author Nikolai Kondrashov <spbnick@gmail.com>
 */

#include "config.h"

#include <libusb.h>
#include <stdlib.h>
#include <stdbool.h>
#include <stdio.h>
#include <strings.h>

#ifndef HAVE_LIBUSB_STRERROR
static const char *
libusb_strerror(enum libusb_error err)
{
    switch (err)
    {
        case LIBUSB_SUCCESS:
            return "Success";
#define MAP(_name, _desc) \
    case LIBUSB_ERROR_##_name:          \
        return _desc " (ERROR_" #_name ")"
	    MAP(IO,
            "Input/output error");
	    MAP(INVALID_PARAM,
            "Invalid parameter");
	    MAP(ACCESS,
            "Access denied (insufficient permissions)");
	    MAP(NO_DEVICE,
            "No such device (it may have been disconnected)");
	    MAP(NOT_FOUND,
            "Entity not found");
	    MAP(BUSY,
            "Resource busy");
	    MAP(TIMEOUT,
            "Operation timed out");
	    MAP(OVERFLOW,
            "Overflow");
	    MAP(PIPE,
            "Pipe error");
	    MAP(INTERRUPTED,
            "System call interrupted (perhaps due to signal)");
	    MAP(NO_MEM,
            "Insufficient memory");
	    MAP(NOT_SUPPORTED,
            "Operation not supported or unimplemented on this platform");
	    MAP(OTHER, "Other error");
#undef MAP
        default:
            return "Unknown error code";
    }
}
#endif

#define GENERIC_ERROR(_fmt, _args...) \
    fprintf(stderr, _fmt "\n", ##_args)

#define GENERIC_FAILURE(_fmt, _args...) \
    GENERIC_ERROR("Failed to " _fmt, ##_args)

#define LIBUSB_FAILURE(_err, _fmt, _args...) \
    GENERIC_FAILURE(_fmt ": %s", ##_args, libusb_strerror(_err))

#define ERROR_CLEANUP(_fmt, _args...) \
    do {                                \
        GENERIC_ERROR(_fmt, ##_args);   \
        goto cleanup;                   \
    } while (0)

#define FAILURE_CLEANUP(_fmt, _args...) \
    do {                                \
        GENERIC_FAILURE(_fmt, ##_args); \
        goto cleanup;                   \
    } while (0)

#define LIBUSB_FAILURE_CLEANUP(_err, _fmt, _args...) \
    do {                                        \
        LIBUSB_FAILURE(_err, _fmt, ##_args);          \
        goto cleanup;                           \
    } while (0)

#define LIBUSB_GUARD(_expr, _fmt, _args...) \
    do {                                                \
        enum libusb_error err = _expr;                  \
        if (err < 0)                                    \
            LIBUSB_FAILURE_CLEANUP(err, _fmt, ##_args); \
    } while (0)

static void
print_chunk(char type, unsigned char *ptr, int len)
{
    printf("%c", type);
    for (; len > 0; ptr++, len--)
        printf(" %.2hhX", *ptr);
    printf("\n");
    fflush(stdout);
}

static int
probe(uint8_t bus_num, uint8_t dev_addr)
{
    int                     result      = 1;
    libusb_context         *ctx         = NULL;
    libusb_device         **dev_list    = NULL;
    ssize_t                 dev_num;
    size_t                  i;
    libusb_device          *dev;
    libusb_device_handle   *handle      = NULL;
    unsigned char           buf[257];
    int                     len;
    uint8_t                 idx_list[]  = {0x64, 0x65, 0x6E, 0x79, 0x7A,
                                           0x7B};
    uint8_t                 idx;

    struct libusb_device_descriptor     dev_desc;

    LIBUSB_GUARD(libusb_init(&ctx), "initialize libusb");

    LIBUSB_GUARD(dev_num = libusb_get_device_list(ctx, &dev_list),
                 "get device list");

    for (i = 0; i < (size_t)dev_num; i++) {
        dev = dev_list[i];
        if (libusb_get_bus_number(dev) == bus_num &&
            libusb_get_device_address(dev) == dev_addr)
            break;
    }
    if (i == (size_t)dev_num)
        ERROR_CLEANUP("Device not found");
    LIBUSB_GUARD(libusb_open(dev, &handle), "open device");
    libusb_free_device_list(dev_list, true);
    dev_list = NULL;

    LIBUSB_GUARD(libusb_get_device_descriptor(dev, &dev_desc),
                 "get device descriptor");
    if (dev_desc.iManufacturer != 0) {
        LIBUSB_GUARD(len = libusb_get_string_descriptor(
                                    handle, dev_desc.iManufacturer,
                                    /* English (United States) */
                                    0x0409,
                                    buf, sizeof(buf)),
                     "get manufacturer string descriptor");
        print_chunk('M', buf + 2, len - 2);
    }
    if (dev_desc.iProduct != 0) {
        LIBUSB_GUARD(len = libusb_get_string_descriptor(
                                    handle, dev_desc.iProduct,
                                    /* English (United States) */
                                    0x0409,
                                    buf, sizeof(buf)),
                     "get product string descriptor");
        print_chunk('P', buf + 2, len - 2);
    }

    for (i = 0; i < sizeof(idx_list) / sizeof(*idx_list); i++)
    {
        idx = idx_list[i];
        buf[0] = idx;

        /* Attempt to get the descriptor */
        len = libusb_get_string_descriptor(
                                handle, idx,
                                /* English (United States) */
                                0x0409,
                                buf + 1, sizeof(buf) - 1);

        /* If the descriptor doesn't exist */
        if (len == LIBUSB_ERROR_PIPE)
            continue;
        LIBUSB_GUARD(len, "get string descriptor 0x%.2X", idx);

        /* Print the descriptor */
        print_chunk('S', buf, len + 1);
    }
    result = 0;

cleanup:

    libusb_close(handle);
    libusb_free_device_list(dev_list, true);
    libusb_exit(ctx);
    return result;
}

static void
usage(FILE *file, const char *name)
{
    fprintf(file,
            "Usage: %s BUS_NUM DEV_ADDR\n"
            "Probe a Huion tablet.\n"
            "\n"
            "Arguments:\n"
            "    BUS_NUM    Bus number.\n"
            "    DEV_ADDR   Device address.\n"
            "\n",
            name);
}

int
main(int argc, char **argv)
{
    const char *name;

    name = rindex(argv[0], '/');
    if (name == NULL)
        name = argv[0];
    else
        name++;

    if (argc != 3) {
        fprintf(stderr, "Invalid number of arguments\n");
        usage(stderr, name);
        exit(1);
    }

    setbuf(stdout, NULL);
    return probe((uint8_t)atoi(argv[1]),
                 (uint8_t)atoi(argv[2]));
}


huion-decode.c
/*
 * Copyright (C) 2013 Nikolai Kondrashov
 *
 * This file is part of huion-tools.
 *
 * Huion-tools is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Huion-tools is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with huion-tools; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 * @author Nikolai Kondrashov <spbnick@gmail.com>
 */

#include <assert.h>
#include <errno.h>
#include <stdint.h>
#include <stdlib.h>
#include <stdio.h>
#include <ctype.h>
#include <strings.h>

#define GENERIC_ERROR(_fmt, _args...) \
    fprintf(stderr, _fmt "\n", ##_args)

#define GENERIC_FAILURE(_fmt, _args...) \
    GENERIC_ERROR("Failed to " _fmt, ##_args)

#define ERROR_CLEANUP(_fmt, _args...) \
    do {                                \
        GENERIC_ERROR(_fmt, ##_args);   \
        goto cleanup;                   \
    } while (0)

#define FAILURE_CLEANUP(_fmt, _args...) \
    do {                                \
        GENERIC_FAILURE(_fmt, ##_args); \
        goto cleanup;                   \
    } while (0)

struct decoder {
    uint8_t     id;
    int       (*decode)(const uint8_t *ptr, int len);
};

static int
decode_params(const uint8_t *ptr, int len)
{
    const int   max_x_off           = 2;
    const int   max_y_off           = 4;
    const int   max_pressure_off    = 8;
    const int   resolution_off      = 10;

#define FIELD(_label, _offset) \
    do {                                                        \
        printf("%14s: ", _label);                               \
        if ((_offset) < len - 1) {                              \
            printf("%u\n",                                      \
                   ptr[_offset] |                               \
                    ((unsigned int)ptr[(_offset) + 1] << 8));   \
        } else {                                                \
            printf("N/A\n");                                    \
        }                                                       \
    } while (0)
    FIELD("Max X", max_x_off);
    FIELD("Max Y", max_y_off);
    FIELD("Max pressure", max_pressure_off);
    FIELD("Resolution", resolution_off);
#undef FIELD
    return 0;
}

static void
print_unicode(const uint8_t *ptr, int len)
{
    uint16_t c;
    for (; len > 1; ptr += 2, len -= 2) {
        c = ptr[0] | (ptr[1] << 8);
        putchar((c >= 0x20 && c <= 0x7E) ? c : '?');
    }
}

static void
print_field_unicode(const char *name, const uint8_t *ptr, int len)
{
    printf("%14s: ", name);
    print_unicode(ptr, len);
    putchar('\n');
}

static int
decode_internal_model(const uint8_t *ptr, int len)
{
    print_field_unicode("Internal model", ptr + 2, len - 2);
    return 0;
}

static int
decode_buttons_status(const uint8_t *ptr, int len)
{
    print_field_unicode("Buttons status", ptr + 2, len - 2);
    return 0;
}

/* List of string descriptor decoders */
static const struct decoder desc_list[] = {
    {0x64, decode_params},
    {0x79, decode_internal_model},
    {0x7b, decode_buttons_status},
    {0x00, NULL}
};

static int
decode_desc(const uint8_t *buf, int len)
{
    const struct decoder   *d;

    if (len == 0) {
        GENERIC_ERROR("String descriptor without index");
        return 1;
    }

    for (d = desc_list; d->decode != NULL; d++) {
        if (d->id == *buf)
            return d->decode(buf + 1, len - 1);
    }

    return 0;
}

static int
decode_manufacturer(const uint8_t *buf, int len)
{
    print_field_unicode("Manufacturer", buf, len);
    return 0;
}

static int
decode_product(const uint8_t *buf, int len)
{
    print_field_unicode("Product", buf, len);
    return 0;
}

/* List of chunk decoders */
static const struct decoder chunk_list[] = {
    {'M',   decode_manufacturer},
    {'P',   decode_product},
    {'S',   decode_desc},
    {'\0',  NULL}
};

static int
decode_chunk(const uint8_t *buf, int len)
{
    const struct decoder   *d;

    assert(len > 0);

    for (d = chunk_list; d->decode != NULL; d++) {
        if (d->id == *buf)
            return d->decode(buf + 1, len - 1);
    }

    return 0;
}

static int
decode(void)
{
    int             result      = 1;
    char           *word        = NULL;
    char           *end;
    uint8_t         buf[256];
    uint8_t        *p           = buf;
    unsigned long   n;

    assert(sizeof(buf) > 0);

    while (scanf("%ms", &word) == 1) {
        /* If it's a single character chunk type */
        if (word[0] != '\0' && !isxdigit(word[0]) && word[1] == '\0') {
            if (p != buf && decode_chunk(buf, p - buf) != 0)
                FAILURE_CLEANUP("decode chunk");
            p = buf;
            n = word[0];
        } else {
            if (p == buf)
                ERROR_CLEANUP("Expecting chunk type indicator");
            errno = 0;
            n = strtoul(word, &end, 16);
            if (*end != '\0' || errno != 0 || n > UINT8_MAX)
                ERROR_CLEANUP("Invalid byte \"%s\"", word);
            if (p >= buf + sizeof(buf))
                ERROR_CLEANUP("Descriptor too long");
        }
        *p++ = n;
        free(word);
        word = NULL;
    };

    if (p != buf && decode_chunk(buf, p - buf) != 0)
        FAILURE_CLEANUP("decode chunk");

    result = 0;

cleanup:
    free(word);
    return result;
}

static void
usage(FILE *file, const char *name)
{
    fprintf(file,
            "Usage: %s\n"
            "Decode a Huion tablet probe dump.\n"
            "\n",
            name);
}

int
main(int argc, char **argv)
{
    const char *name;

    name = rindex(argv[0], '/');
    if (name == NULL)
        name = argv[0];
    else
        name++;

    if (argc != 1) {
        fprintf(stderr, "Invalid number of arguments\n");
        usage(stderr, name);
        exit(1);
    }

    setlinebuf(stdout);
    return decode();
}


uclogic-decode.c
/*
 * Copyright (C) 2013 Nikolai Kondrashov
 *
 * This file is part of uclogic-tools.
 *
 * Uclogic-tools is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Uclogic-tools is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with uclogic-tools; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 * @author Nikolai Kondrashov <spbnick@gmail.com>
 */

#include <assert.h>
#include <errno.h>
#include <stdint.h>
#include <stdlib.h>
#include <stdio.h>
#include <ctype.h>
#include <string.h>
#include <strings.h>

#define GENERIC_ERROR(_fmt, _args...) \
    fprintf(stderr, _fmt "\n", ##_args)

#define GENERIC_FAILURE(_fmt, _args...) \
    GENERIC_ERROR("Failed to " _fmt, ##_args)

#define ERROR_CLEANUP(_fmt, _args...) \
    do {                                \
        GENERIC_ERROR(_fmt, ##_args);   \
        goto cleanup;                   \
    } while (0)

#define FAILURE_CLEANUP(_fmt, _args...) \
    do {                                \
        GENERIC_FAILURE(_fmt, ##_args); \
        goto cleanup;                   \
    } while (0)

/** Format string for a field header */
#define FIELD_HEAD_FMT "    | %22s: "

/** Format string for a field header with hex index */
#define IDX_FIELD_HEAD_FMT " %02x | %22s: "

/** Format string for a sub-field header */
#define SUB_FIELD_HEAD_FMT "    | %30s: "

/** Blank output line string */
#define BLANK_LINE_STR "    |"

struct decoder {
    /** Decoder ID */
    uint8_t     id;
    /**
     * Decoder function.
     *
     * @param idx   Index of the decoded data (e.g. descriptor number).
     * @param ptr   Pointer to the decoded data.
     * @param len   Length of the decoded data.
     *
     * @return Zero if decoded successfully, non-zero otherwise.
     */
    int       (*decode)(uint8_t idx, const uint8_t *ptr, int len);
};

/**
 * Print a 16-bit unsigned sub-field.
 * Assumes "ptr" and "len" variables are in scope.
 *
 * @param _offset   Field offset from ptr, bytes.
 * @param _label    Field label string literal.
 */
#define PRINT_SUB_FIELD_U16(_offset, _label) \
    do {                                                        \
        printf(SUB_FIELD_HEAD_FMT, _label);                     \
        if ((_offset) + 1 < len) {                              \
            printf("%u\n",                                      \
                   ptr[_offset] |                               \
                    ((unsigned int)ptr[(_offset) + 1] << 8));   \
        } else {                                                \
            printf("N/A\n");                                    \
        }                                                       \
    } while (0)

/**
 * Print a 24-bit unsigned sub-field.
 * Assumes "ptr" and "len" variables are in scope.
 *
 * @param _offset   Field offset from ptr, bytes.
 * @param _label    Field label string literal.
 */
#define PRINT_SUB_FIELD_U24(_offset, _label) \
    do {                                                        \
        printf(SUB_FIELD_HEAD_FMT, _label);                     \
        if ((_offset) + 2 < len) {                              \
            printf("%u\n",                                      \
                   ptr[_offset] |                               \
                    ((unsigned int)ptr[(_offset) + 1] << 8) |   \
                    ((unsigned int)ptr[(_offset) + 2] << 16));  \
        } else {                                                \
            printf("N/A\n");                                    \
        }                                                       \
    } while (0)

static void
print_unicode(const uint8_t *ptr, int len)
{
    uint16_t c;
    for (; len > 1; ptr += 2, len -= 2) {
        c = ptr[0] | (ptr[1] << 8);
        putchar((c >= 0x20 && c <= 0x7E) ? c : '?');
    }
}

static void
print_field_unicode(const char *name, const uint8_t *ptr, int len)
{
    printf(FIELD_HEAD_FMT, name);
    print_unicode(ptr, len);
    putchar('\n');
}

static void
print_idx_field_unicode(uint8_t idx, const char *name,
                        const uint8_t *ptr, int len)
{
    printf(IDX_FIELD_HEAD_FMT, idx, name);
    print_unicode(ptr, len);
    putchar('\n');
}

static int
decode_params1(uint8_t idx, const uint8_t *ptr, int len)
{
    print_idx_field_unicode(idx, "Params block #1", ptr + 2, len - 2);
    PRINT_SUB_FIELD_U16(2, "Max X");
    PRINT_SUB_FIELD_U16(4, "Max Y");
    PRINT_SUB_FIELD_U16(8, "Max pressure");
    PRINT_SUB_FIELD_U16(10, "Resolution");
    puts(BLANK_LINE_STR);
    return 0;
}

static int
decode_internal_model(uint8_t idx, const uint8_t *ptr, int len)
{
    print_idx_field_unicode(idx, "Internal model", ptr + 2, len - 2);
    return 0;
}

static int
decode_buttons_status(uint8_t idx, const uint8_t *ptr, int len)
{
    print_idx_field_unicode(idx, "Buttons status", ptr + 2, len - 2);
    return 0;
}

static int
decode_params2(uint8_t idx, const uint8_t *ptr, int len)
{
    print_idx_field_unicode(idx, "Params block #2", ptr + 2, len - 2);
    PRINT_SUB_FIELD_U24(2, "Max X");
    PRINT_SUB_FIELD_U24(5, "Max Y");
    PRINT_SUB_FIELD_U16(8, "Max pressure");
    PRINT_SUB_FIELD_U16(10, "Resolution");
    puts(BLANK_LINE_STR);
    return 0;
}

static int
decode_firmware_version(uint8_t idx, const uint8_t *ptr, int len)
{
    print_idx_field_unicode(idx, "Firmware version", ptr + 2, len - 2);
    return 0;
}

static int
decode_internal_manufacturer(uint8_t idx, const uint8_t *ptr, int len)
{
    print_idx_field_unicode(idx, "Internal manufacturer", ptr + 2, len - 2);
    return 0;
}

/* List of string descriptor decoders */
static const struct decoder desc_list[] = {
    {0x64, decode_params1},
    {0x79, decode_internal_model},
    {0x7b, decode_buttons_status},
    {0xc8, decode_params2},
    {0xc9, decode_firmware_version},
    {0xca, decode_internal_manufacturer},
    {0x00, NULL}
};

static int
decode_desc(uint8_t idx, const uint8_t *buf, int len)
{
    const struct decoder   *d;

    (void)idx;

    if (len == 0) {
        GENERIC_ERROR("String descriptor without index");
        return 1;
    }

    for (d = desc_list; d->decode != NULL; d++) {
        if (d->id == *buf)
            return d->decode(d->id, buf + 1, len - 1);
    }

    return 0;
}

static int
decode_manufacturer(uint8_t idx, const uint8_t *buf, int len)
{
    (void)idx;
    print_field_unicode("Manufacturer", buf, len);
    return 0;
}

static int
decode_product(uint8_t idx, const uint8_t *buf, int len)
{
    (void)idx;
    print_field_unicode("Product", buf, len);
    return 0;
}

/* List of chunk decoders */
static const struct decoder chunk_list[] = {
    {'M',   decode_manufacturer},
    {'P',   decode_product},
    {'S',   decode_desc},
    {'\0',  NULL}
};

static int
decode_chunk(const uint8_t *buf, int len)
{
    const struct decoder   *d;

    assert(len > 0);

    for (d = chunk_list; d->decode != NULL; d++) {
        if (d->id == *buf)
            return d->decode(d->id, buf + 1, len - 1);
    }

    return 0;
}

static int
decode(FILE *input)
{
    int             result      = 1;
    char           *word        = NULL;
    char           *end;
    /*
     * Chunk type byte +
     * string descriptor index byte +
     * maximum descriptor length
     */
    uint8_t         buf[258];
    uint8_t        *p           = buf;
    unsigned long   n;

    assert(sizeof(buf) > 0);

    /* For each "word" (a run of non-whitespace characters) on stdin */
    while (fscanf(input, "%ms", &word) == 1) {
        /* If it's a single character chunk type */
        if (word[0] != '\0' && !isxdigit(word[0]) && word[1] == '\0') {
            /* If we read a chunk before and it is failing decoding */
            if (p != buf && decode_chunk(buf, p - buf) != 0)
                FAILURE_CLEANUP("decode chunk");
            /* Start a new chunk */
            p = buf;
            n = word[0];
        /* Otherwise it is some other word */
        } else {
            /* If there was nothing else for this chunk (i.e. no type) */
            if (p == buf)
                ERROR_CLEANUP("Expecting chunk type indicator");
            /* Try to decode the word as a hex byte */
            errno = 0;
            n = strtoul(word, &end, 16);
            if (*end != '\0' || errno != 0 || n > UINT8_MAX)
                ERROR_CLEANUP("Invalid byte \"%s\"", word);
            /* If there's no space in the buffer */
            if (p >= buf + sizeof(buf))
                ERROR_CLEANUP("Descriptor too long");
        }
        /* Write the byte into the buffer */
        *p++ = n;
        /* Discard the read word */
        free(word);
        word = NULL;
    };

    if (p != buf && decode_chunk(buf, p - buf) != 0)
        FAILURE_CLEANUP("decode chunk");

    result = 0;

cleanup:
    free(word);
    return result;
}

static void
usage(FILE *file, const char *name)
{
    fprintf(file,
            "Usage: %s [PROBE_OUTPUT]\n"
            "Decode a UC-Logic tablet probe dump.\n"
            "\n",
            name);
}

int
main(int argc, char **argv)
{
    const char *name;
    FILE *input;

    name = rindex(argv[0], '/');
    if (name == NULL)
        name = argv[0];
    else
        name++;

    if (argc > 2) {
        fprintf(stderr, "Invalid number of arguments\n");
        usage(stderr, name);
        exit(1);
    } else if (argc > 1) {
        input = fopen(argv[1], "r");
        if (input == NULL) {
            fprintf(stderr, "Failed opening %s: %s\n", argv[1],
                    strerror(errno));
        }
    } else {
        input = stdin;
    }

    setlinebuf(stdout);
    return decode(input);
}
